# Data Pre-Processing

**The original dataset can be found on Kaggle**

**URL = https://www.kaggle.com/datasets/trolukovich/steam-games-complete-dataset?resource=download**

**It is licensed with "CC0: Public Domain" which allows me to use the dataset**

# 1. Loading the dataset
**Starting with importing libraries and reading the dataset.**

In [1]:
import re
import warnings

import pandas as pd

from collections import Counter

warnings.filterwarnings("ignore")

In [2]:
dataframe = pd.read_csv("Datasets/steam_games.csv")
dataframe.shape

(40833, 20)

In [3]:
dataframe.head()

url types   
0    https://store.steampowered.com/app/379720/DOOM/   app  \
1  https://store.steampowered.com/app/578080/PLAY...   app   
2  https://store.steampowered.com/app/637090/BATT...   app   
3    https://store.steampowered.com/app/221100/DayZ/   app   
4  https://store.steampowered.com/app/8500/EVE_On...   app   

                            name   
0                           DOOM  \
1  PLAYERUNKNOWN'S BATTLEGROUNDS   
2                     BATTLETECH   
3                           DayZ   
4                     EVE Online   

                                        desc_snippet   
0  Now includes all three premium DLC packs (Unto...  \
1  PLAYERUNKNOWN'S BATTLEGROUNDS is a battle roya...   
2  Take command of your own mercenary outfit of '...   
3  The post-soviet country of Chernarus is struck...   
4  EVE Online is a community-driven spaceship MMO...   

                                      recent_reviews   
0  Very Positive,(554),- 89% of the 554 user revi...  \
1  Mixed,(6,214),- 49% of the 6,214 user reviews ...   
2  Mixed,(166),- 54% of the 166 user reviews in t...   
3  Mixed,(932),- 57% of the 932 user reviews in t...   
4  Mixed,(287),- 54% of the 287 user reviews in t...   

                                         all_reviews  release_date   
0  Very Positive,(42,550),- 92% of the 42,550 use...  May 12, 2016  \
1  Mixed,(836,608),- 49% of the 836,608 user revi...  Dec 21, 2017   
2  Mostly Positive,(7,030),- 71% of the 7,030 use...  Apr 24, 2018   
3  Mixed,(167,115),- 61% of the 167,115 user revi...  Dec 13, 2018   
4  Mostly Positive,(11,481),- 74% of the 11,481 u...   May 6, 2003   

             developer                                publisher   
0          id Software    Bethesda Softworks,Bethesda Softworks  \
1     PUBG Corporation        PUBG Corporation,PUBG Corporation   
2  Harebrained Schemes  Paradox Interactive,Paradox Interactive   
3  Bohemia Interactive  Bohemia Interactive,Bohemia Interactive   
4                  CCP                                  CCP,CCP   

                                        popular_tags   
0  FPS,Gore,Action,Demons,Shooter,First-Person,Gr...  \
1  Survival,Shooter,Multiplayer,Battle Royale,PvP...   
2  Mechs,Strategy,Turn-Based,Turn-Based Tactics,S...   
3  Survival,Zombies,Open World,Multiplayer,PvP,Ma...   
4  Space,Massively Multiplayer,Sci-fi,Sandbox,MMO...   

                                        game_details   
0  Single-player,Multi-player,Co-op,Steam Achieve...  \
1             Multi-player,Online Multi-Player,Stats   
2  Single-player,Multi-player,Online Multi-Player...   
3  Multi-player,Online Multi-Player,Steam Worksho...   
4  Multi-player,Online Multi-Player,MMO,Co-op,Onl...   

                                           languages  achievements   
0  English,French,Italian,German,Spanish - Spain,...          54.0  \
1  English,Korean,Simplified Chinese,French,Germa...          37.0   
2                      English,French,German,Russian         128.0   
3  English,French,Italian,German,Spanish - Spain,...           NaN   
4                      English,German,Russian,French           NaN   

                                               genre   
0                                             Action  \
1             Action,Adventure,Massively Multiplayer   
2                          Action,Adventure,Strategy   
3             Action,Adventure,Massively Multiplayer   
4  Action,Free to Play,Massively Multiplayer,RPG,...   

                                    game_description   
0   About This Game Developed by id software, the...  \
1   About This Game  PLAYERUNKNOWN'S BATTLEGROUND...   
2   About This Game  From original BATTLETECH/Mec...   
3   About This Game The post-soviet country of Ch...   
4                                  About This Game     

                                      mature_content   
0                                                NaN  \
1   Mature Content Description  The developers de...   
2                       

# 2. Cleaning the dataset
**I get the idea how data looks now. Next step is to check for missing values in dataset and drop rows with missing one.**

**1. Column "name"**

In [4]:
dataframe.loc[dataframe["name"].isna()].index

Index([  704,  4847,  6381,  7869,  9615,  9616,  9956, 12146, 12879, 23099,
       28380, 31321, 34989, 34991, 35169, 39575],
      dtype='int64')

In [5]:
dataframe = dataframe.drop(dataframe.loc[dataframe["name"].isna()].index)
dataframe.loc[dataframe["name"].isna()].index

Index([], dtype='int64')

In [6]:
dataframe.shape

(40817, 20)

**I have 16 rows less now.**

**2. Irrelevant columns**

**Next step is to drop columns which provide us with no usable data. Taking into account the goal - predict the popularity of the new game. Upon release most of this data is absent.**

In [7]:
dataframe_clean = dataframe.drop(["url", "types", "desc_snippet", "recent_reviews", "achievements",
                                      "game_description", "mature_content", "minimum_requirements",
                                     "recommended_requirements", "original_price", "discount_price"], axis=1)
dataframe_clean.head()

name   
0                           DOOM  \
1  PLAYERUNKNOWN'S BATTLEGROUNDS   
2                     BATTLETECH   
3                           DayZ   
4                     EVE Online   

                                         all_reviews  release_date   
0  Very Positive,(42,550),- 92% of the 42,550 use...  May 12, 2016  \
1  Mixed,(836,608),- 49% of the 836,608 user revi...  Dec 21, 2017   
2  Mostly Positive,(7,030),- 71% of the 7,030 use...  Apr 24, 2018   
3  Mixed,(167,115),- 61% of the 167,115 user revi...  Dec 13, 2018   
4  Mostly Positive,(11,481),- 74% of the 11,481 u...   May 6, 2003   

             developer                                publisher   
0          id Software    Bethesda Softworks,Bethesda Softworks  \
1     PUBG Corporation        PUBG Corporation,PUBG Corporation   
2  Harebrained Schemes  Paradox Interactive,Paradox Interactive   
3  Bohemia Interactive  Bohemia Interactive,Bohemia Interactive   
4                  CCP                                  CCP,CCP   

                                        popular_tags   
0  FPS,Gore,Action,Demons,Shooter,First-Person,Gr...  \
1  Survival,Shooter,Multiplayer,Battle Royale,PvP...   
2  Mechs,Strategy,Turn-Based,Turn-Based Tactics,S...   
3  Survival,Zombies,Open World,Multiplayer,PvP,Ma...   
4  Space,Massively Multiplayer,Sci-fi,Sandbox,MMO...   

                                        game_details   
0  Single-player,Multi-player,Co-op,Steam Achieve...  \
1             Multi-player,Online Multi-Player,Stats   
2  Single-player,Multi-player,Online Multi-Player...   
3  Multi-player,Online Multi-Player,Steam Worksho...   
4  Multi-player,Online Multi-Player,MMO,Co-op,Onl...   

                                           languages   
0  English,French,Italian,German,Spanish - Spain,...  \
1  English,Korean,Simplified Chinese,French,Germa...   
2                      English,French,German,Russian   
3  English,French,Italian,German,Spanish - Spain,...   
4                      English,German,Russian,French   

                                               genre  
0                                             Action  
1             Action,Adventure,Massively Multiplayer  
2                          Action,Adventure,Strategy  
3             Action,Adventure,Massively Multiplayer  
4  Action,Free to Play,Massively Multiplayer,RPG,...

**3. Column "all_reviews"**

In [8]:
dataframe_clean.loc[dataframe_clean["all_reviews"].isna()].index

Index([    5,    15,    25,    39,    44,    57,    59,    63,    66,    68,
       ...
       40823, 40824, 40825, 40826, 40827, 40828, 40829, 40830, 40831, 40832],
      dtype='int64', length=12347)

In [9]:
dataframe_clean = dataframe_clean.drop(dataframe_clean.loc[dataframe_clean["all_reviews"].isna()].index)
dataframe_clean.loc[dataframe_clean["all_reviews"].isna()].index

Index([], dtype='int64')

In [10]:
dataframe_clean.shape

(28470, 9)

**12347 rows less. Sadly, this is the column from where I will be extracting my target data and it is bad idea to fill the missing values as it may affect the model predictions.**

**4. Column "release_date"**

In [11]:
dataframe_clean.loc[dataframe_clean["release_date"].isna()].index

Index([  227,   329,   847,   849,  1039,  1435,  1631,  3280,  3422,  4758,
       ...
       28995, 29243, 29854, 29884, 30014, 30015, 31154, 33952, 34013, 34057],
      dtype='int64', length=132)

In [12]:
dataframe_clean = dataframe_clean.drop(dataframe_clean.loc[dataframe_clean["release_date"].isna()].index)
dataframe_clean.loc[dataframe_clean["release_date"].isna()].index

Index([], dtype='int64')

In [13]:
dataframe_clean.shape

(28338, 9)

**132 less rows. Insignificant on the chunk of data left.**

**5. Column "developer"**

In [14]:
dataframe_clean.loc[dataframe_clean["developer"].isna()].index

Index([  146,  1101,  1854,  2517,  2712,  3471,  3989,  4722,  4854,  4933,
       ...
       30042, 30246, 30248, 30293, 30608, 31152, 31170, 31231, 34058, 34304],
      dtype='int64', length=155)

In [15]:
dataframe_clean = dataframe_clean.drop(dataframe_clean.loc[dataframe_clean["developer"].isna()].index)
dataframe_clean.loc[dataframe_clean["developer"].isna()].index

Index([], dtype='int64')

In [16]:
dataframe_clean.shape

(28183, 9)

**155 less rows. Insignificant on the chunk of data left.**

**6. Column "publisher"**

In [17]:
dataframe_clean.loc[dataframe_clean["publisher"].isna()].index

Index([   45,  1061,  1899,  1979,  2422,  2574,  2645,  2711,  2854,  2941,
       ...
       34390, 34529, 34610, 34732, 35581, 36090, 37946, 37947, 37948, 38360],
      dtype='int64', length=2096)

In [18]:
dataframe_clean = dataframe_clean.drop(dataframe_clean.loc[dataframe_clean["publisher"].isna()].index)
dataframe_clean.loc[dataframe_clean["publisher"].isna()].index

Index([], dtype='int64')

In [19]:
dataframe_clean.shape

(26087, 9)

**2096 less rows. Significant loss, but no way to fill the missing data.**

**7. Column "popular_tags"**

In [20]:
dataframe_clean.loc[dataframe_clean["popular_tags"].isna()].index

Index([ 6027, 10649, 10664, 11206, 11377, 11379, 11814, 11940, 11942, 12508,
       12663, 13206, 13213, 13967, 14104, 14702, 14763, 15580, 15635, 15637,
       15740, 16082, 16139, 16583, 17437, 18580, 19396, 19996, 20268, 21321,
       21642, 22442, 23009, 23050, 23414, 23562, 24129, 25036, 27069, 27845,
       28949, 31218, 31615, 32041, 32045, 34271, 34608],
      dtype='int64')

In [21]:
dataframe_clean = dataframe_clean.drop(dataframe_clean.loc[dataframe_clean["popular_tags"].isna()].index)
dataframe_clean.loc[dataframe_clean["popular_tags"].isna()].index

Index([], dtype='int64')

In [22]:
dataframe_clean.shape

(26040, 9)

**47 less rows. Insignificant on the chunk of data left.**

**8. Column "game_details"**

In [23]:
dataframe_clean.loc[dataframe_clean["game_details"].isna()].index

Index([  478,   663,   727,   859,   949,  1007,  1052,  1114,  1153,  1305,
       ...
       30623, 30645, 30669, 30758, 30787, 30821, 35821, 36154, 36821, 37456],
      dtype='int64', length=273)

In [24]:
dataframe_clean = dataframe_clean.drop(dataframe_clean.loc[dataframe_clean["game_details"].isna()].index)
dataframe_clean.loc[dataframe_clean["game_details"].isna()].index

Index([], dtype='int64')

In [25]:
dataframe_clean.shape

(25767, 9)

**273 less rows. Insignificant on the chunk of data left.**

**9. Column "languages"**

In [26]:
dataframe_clean.loc[dataframe_clean["languages"].isna()].index

Index([], dtype='int64')

**No missing data.**

**10. Column "genre"**

In [27]:
dataframe_clean.loc[dataframe_clean["genre"].isna()].index

Index([  528,  1613,  2254,  3639,  4018,  5211,  6127,  6743,  6990,  7138,
        7510,  7836,  8239,  8780, 10677, 11948, 12388, 12987, 13857, 14871,
       14879, 14880, 14884, 15119, 16370, 18200, 18268, 19055, 21998, 22570,
       23695, 23696, 24331, 24884, 26416, 27277, 27402, 29379],
      dtype='int64')

In [28]:
dataframe_clean = dataframe_clean.drop(dataframe_clean.loc[dataframe_clean["genre"].isna()].index)
dataframe_clean.loc[dataframe_clean["genre"].isna()].index

Index([], dtype='int64')

In [29]:
dataframe_clean.shape

(25729, 9)

**38 less rows. Insignificant on the chunk of data left.**

# 3. Reshaping the dataset

**1. Creating the column "has_setting". It describes wether the game starts the setting or continue it.**

In [30]:
# function looks for numbers in range (2-10) in the column "name" and returns True if there is a number in a string
def check_for_setting(name):
    setting = re.findall(r"\d+", name)
    if setting:
        for i in setting:
            if 2 <= int(i) <= 10:
                return True
    return False

In [31]:
dataframe_clean["has_setting"] = dataframe_clean["name"].apply(check_for_setting)
dataframe_clean = dataframe_clean.set_index("name")

In [32]:
has_setting = dataframe_clean[dataframe_clean["has_setting"]]
has_setting.shape

(2669, 9)

**2669 rows in dataset respond with True for a column "has_setting"**

**2. Changing the datatype in column "release_date" to an int which indicates the year of releas.**

In [33]:
dataframe_clean["release_date"] = pd.to_datetime(dataframe_clean["release_date"], 
                                                 format="%b %d, %Y", errors="coerce").dt.year.astype("Int64")

**Checking if some data was lost in a process and deleting rows where it is absent.**

In [34]:
dataframe_clean.loc[dataframe_clean["release_date"].isna()].index

Index(['Aura Kingdom', 'Technobabylon', 'Age of Wonders Shadow Magic',
       'Puzzle Agent', 'NECROPOLIS: BRUTAL EDITION', 'Red Stone Online',
       'Silent Hunter 5®: Battle of the Atlantic',
       'Sea Dogs: To Each His Own - Pirate Open World RPG', 'Age of Wonders',
       'Idle Racing GO: Clicker Tycoon', 'Muffin Knight',
       'NOBUNAGA'S AMBITION: Souzou (Traditional Chinese version)',
       'DW8XLCE - JAPANESE VOICE OPTION',
       'Nobunaga's Ambition: Souzou - Nobunaga Oda In-Game Face CG',
       'Tetrapulse',
       'Nobunaga's Ambition: Souzou - Series 30th Anniversary Contents',
       'Silent Escape: Induction', 'Apparition', 'Smith and Winston',
       'Aliens vs Predator™ Bughunt Map Pack', 'Depression The Game',
       'Age of Wonders II: The Wizard's Throne', 'Altitude0: Lower & Faster',
       'Turba', 'Call of Duty®: Advanced Warfare - Season Pass',
       'NIGHTSTAR: Rogue Wings', 'Cirque du Soleil',
       '3DMark Time Spy benchmark', 'Darkest of Days', 'Blue

In [35]:
dataframe_clean = dataframe_clean.drop(dataframe_clean.loc[dataframe_clean["release_date"].isna()].index)
dataframe_clean.loc[dataframe_clean["release_date"].isna()].index

Index([], dtype='object', name='name')

In [36]:
dataframe_clean.head()

all_reviews   
name                                                                               
DOOM                           Very Positive,(42,550),- 92% of the 42,550 use...  \
PLAYERUNKNOWN'S BATTLEGROUNDS  Mixed,(836,608),- 49% of the 836,608 user revi...   
BATTLETECH                     Mostly Positive,(7,030),- 71% of the 7,030 use...   
DayZ                           Mixed,(167,115),- 61% of the 167,115 user revi...   
EVE Online                     Mostly Positive,(11,481),- 74% of the 11,481 u...   

                               release_date            developer   
name                                                               
DOOM                                   2016          id Software  \
PLAYERUNKNOWN'S BATTLEGROUNDS          2017     PUBG Corporation   
BATTLETECH                             2018  Harebrained Schemes   
DayZ                                   2018  Bohemia Interactive   
EVE Online                             2003                  CCP   

                                                             publisher   
name                                                                     
DOOM                             Bethesda Softworks,Bethesda Softworks  \
PLAYERUNKNOWN'S BATTLEGROUNDS        PUBG Corporation,PUBG Corporation   
BATTLETECH                     Paradox Interactive,Paradox Interactive   
DayZ                           Bohemia Interactive,Bohemia Interactive   
EVE Online                                                     CCP,CCP   

                                                                    popular_tags   
name                                                                               
DOOM                           FPS,Gore,Action,Demons,Shooter,First-Person,Gr...  \
PLAYERUNKNOWN'S BATTLEGROUNDS  Survival,Shooter,Multiplayer,Battle Royale,PvP...   
BATTLETECH                     Mechs,Strategy,Turn-Based,Turn-Based Tactics,S...   
DayZ                           Survival,Zombies,Open World,Multiplayer,PvP,Ma...   
EVE Online                     Space,Massively Multiplayer,Sci-fi,Sandbox,MMO...   

                                                                    game_details   
name                                                                               
DOOM                           Single-player,Multi-player,Co-op,Steam Achieve...  \
PLAYERUNKNOWN'S BATTLEGROUNDS             Multi-player,Online Multi-Player,Stats   
BATTLETECH                     Single-player,Multi-player,Online Multi-Player...   
DayZ                           Multi-player,Online Multi-Player,Steam Worksho...   
EVE Online                     Multi-player,Online Multi-Player,MMO,Co-op,Onl...   

                                                                       languages   
name                                                                               
DOOM                           English,French,Italian,German,Spanish - Spain,...  \
PLAYERUNKNOWN'S BATTLEGROUNDS  English,Korean,Simplified Chinese,French,Germa...   
BATTLETECH                                         English,French,German,Russian   
DayZ                           English,French,Italian,German,Spanish - Spain,...   
EVE Online                                         English,German,Russian,French   

                                                                           genre   
name                                                                               
DOOM                                                                      Action  \
PLAYERUNKNOWN'S BATTLEGROUNDS             Action,Adventure,Massively Multiplayer   
BATTLETECH                                             Action,Adventure,Strategy   
DayZ                                      Action,Adventure,Massively Multiplayer   
EVE Online                     Action,Free to Play,Massively Multiplayer,RPG,...   

                               has_setting  
name                                        
DOOM                                 F

**33 rows were deleted in a process.**

**3. Creating the column "published_by_developer". The column will have a boolean type of data regarding the data in columns "publisher" and "developer".**

**removing the duplicated strings in column "publisher".**

In [37]:
# function spilts the string and returns unuiqe part of it
def remove_duplicates(string):
    publishers = string.split(",")
    unique_publishers = list(set(publishers))
    return ",".join(unique_publishers)

In [38]:
dataframe_clean["publisher"] = dataframe_clean["publisher"].apply(remove_duplicates)

In [39]:
dataframe_clean.head()

all_reviews   
name                                                                               
DOOM                           Very Positive,(42,550),- 92% of the 42,550 use...  \
PLAYERUNKNOWN'S BATTLEGROUNDS  Mixed,(836,608),- 49% of the 836,608 user revi...   
BATTLETECH                     Mostly Positive,(7,030),- 71% of the 7,030 use...   
DayZ                           Mixed,(167,115),- 61% of the 167,115 user revi...   
EVE Online                     Mostly Positive,(11,481),- 74% of the 11,481 u...   

                               release_date            developer   
name                                                               
DOOM                                   2016          id Software  \
PLAYERUNKNOWN'S BATTLEGROUNDS          2017     PUBG Corporation   
BATTLETECH                             2018  Harebrained Schemes   
DayZ                                   2018  Bohemia Interactive   
EVE Online                             2003                  CCP   

                                         publisher   
name                                                 
DOOM                            Bethesda Softworks  \
PLAYERUNKNOWN'S BATTLEGROUNDS     PUBG Corporation   
BATTLETECH                     Paradox Interactive   
DayZ                           Bohemia Interactive   
EVE Online                                     CCP   

                                                                    popular_tags   
name                                                                               
DOOM                           FPS,Gore,Action,Demons,Shooter,First-Person,Gr...  \
PLAYERUNKNOWN'S BATTLEGROUNDS  Survival,Shooter,Multiplayer,Battle Royale,PvP...   
BATTLETECH                     Mechs,Strategy,Turn-Based,Turn-Based Tactics,S...   
DayZ                           Survival,Zombies,Open World,Multiplayer,PvP,Ma...   
EVE Online                     Space,Massively Multiplayer,Sci-fi,Sandbox,MMO...   

                                                                    game_details   
name                                                                               
DOOM                           Single-player,Multi-player,Co-op,Steam Achieve...  \
PLAYERUNKNOWN'S BATTLEGROUNDS             Multi-player,Online Multi-Player,Stats   
BATTLETECH                     Single-player,Multi-player,Online Multi-Player...   
DayZ                           Multi-player,Online Multi-Player,Steam Worksho...   
EVE Online                     Multi-player,Online Multi-Player,MMO,Co-op,Onl...   

                                                                       languages   
name                                                                               
DOOM                           English,French,Italian,German,Spanish - Spain,...  \
PLAYERUNKNOWN'S BATTLEGROUNDS  English,Korean,Simplified Chinese,French,Germa...   
BATTLETECH                                         English,French,German,Russian   
DayZ                           English,French,Italian,German,Spanish - Spain,...   
EVE Online                                         English,German,Russian,French   

                                                                           genre   
name                                                                               
DOOM                                                                      Action  \
PLAYERUNKNOWN'S BATTLEGROUNDS             Action,Adventure,Massively Multiplayer   
BATTLETECH                                             Action,Adventure,Strategy   
DayZ                                      Action,Adventure,Massively Multiplayer   
EVE Online                     Action,Free to Play,Massively Multiplayer,RPG,...   

                               has_setting  
name                                        
DOOM                                 False  
PLAYERUNKNOWN'S BATTLEGROUNDS        False  
BATTLETECH                           False  
DayZ                                 False 

**Creating the column "published_by_developer" and deliting columns "developer" and "publisher".**

In [40]:
dataframe_clean["published_by_developer"] = dataframe_clean["developer"] == dataframe_clean["publisher"]

In [41]:
published_by_developer = dataframe_clean[dataframe_clean["published_by_developer"]]
dataframe_clean = dataframe_clean.drop(["developer", "publisher"], axis=1)
dataframe_clean.head()

all_reviews   
name                                                                               
DOOM                           Very Positive,(42,550),- 92% of the 42,550 use...  \
PLAYERUNKNOWN'S BATTLEGROUNDS  Mixed,(836,608),- 49% of the 836,608 user revi...   
BATTLETECH                     Mostly Positive,(7,030),- 71% of the 7,030 use...   
DayZ                           Mixed,(167,115),- 61% of the 167,115 user revi...   
EVE Online                     Mostly Positive,(11,481),- 74% of the 11,481 u...   

                               release_date   
name                                          
DOOM                                   2016  \
PLAYERUNKNOWN'S BATTLEGROUNDS          2017   
BATTLETECH                             2018   
DayZ                                   2018   
EVE Online                             2003   

                                                                    popular_tags   
name                                                                               
DOOM                           FPS,Gore,Action,Demons,Shooter,First-Person,Gr...  \
PLAYERUNKNOWN'S BATTLEGROUNDS  Survival,Shooter,Multiplayer,Battle Royale,PvP...   
BATTLETECH                     Mechs,Strategy,Turn-Based,Turn-Based Tactics,S...   
DayZ                           Survival,Zombies,Open World,Multiplayer,PvP,Ma...   
EVE Online                     Space,Massively Multiplayer,Sci-fi,Sandbox,MMO...   

                                                                    game_details   
name                                                                               
DOOM                           Single-player,Multi-player,Co-op,Steam Achieve...  \
PLAYERUNKNOWN'S BATTLEGROUNDS             Multi-player,Online Multi-Player,Stats   
BATTLETECH                     Single-player,Multi-player,Online Multi-Player...   
DayZ                           Multi-player,Online Multi-Player,Steam Worksho...   
EVE Online                     Multi-player,Online Multi-Player,MMO,Co-op,Onl...   

                                                                       languages   
name                                                                               
DOOM                           English,French,Italian,German,Spanish - Spain,...  \
PLAYERUNKNOWN'S BATTLEGROUNDS  English,Korean,Simplified Chinese,French,Germa...   
BATTLETECH                                         English,French,German,Russian   
DayZ                           English,French,Italian,German,Spanish - Spain,...   
EVE Online                                         English,German,Russian,French   

                                                                           genre   
name                                                                               
DOOM                                                                      Action  \
PLAYERUNKNOWN'S BATTLEGROUNDS             Action,Adventure,Massively Multiplayer   
BATTLETECH                                             Action,Adventure,Strategy   
DayZ                                      Action,Adventure,Massively Multiplayer   
EVE Online                     Action,Free to Play,Massively Multiplayer,RPG,...   

                               has_setting  published_by_developer  
name                                                                
DOOM                                 False                   False  
PLAYERUNKNOWN'S BATTLEGROUNDS        False                    True  
BATTLETECH                           False                   False  
DayZ                                 False                    True  
EVE Online                           False                    True

**4. Creating the column "multiple_languages". The column will have a boolean type of data regarding the data in column "languages". Drop the cloumn "languages" after data extraction.**

In [42]:
dataframe_clean["multiple_languages"] = dataframe_clean["languages"].str.split(",").apply(lambda x: len(x) > 1)
dataframe_clean = dataframe_clean.drop(["languages"], axis=1)
dataframe_clean.head()

all_reviews   
name                                                                               
DOOM                           Very Positive,(42,550),- 92% of the 42,550 use...  \
PLAYERUNKNOWN'S BATTLEGROUNDS  Mixed,(836,608),- 49% of the 836,608 user revi...   
BATTLETECH                     Mostly Positive,(7,030),- 71% of the 7,030 use...   
DayZ                           Mixed,(167,115),- 61% of the 167,115 user revi...   
EVE Online                     Mostly Positive,(11,481),- 74% of the 11,481 u...   

                               release_date   
name                                          
DOOM                                   2016  \
PLAYERUNKNOWN'S BATTLEGROUNDS          2017   
BATTLETECH                             2018   
DayZ                                   2018   
EVE Online                             2003   

                                                                    popular_tags   
name                                                                               
DOOM                           FPS,Gore,Action,Demons,Shooter,First-Person,Gr...  \
PLAYERUNKNOWN'S BATTLEGROUNDS  Survival,Shooter,Multiplayer,Battle Royale,PvP...   
BATTLETECH                     Mechs,Strategy,Turn-Based,Turn-Based Tactics,S...   
DayZ                           Survival,Zombies,Open World,Multiplayer,PvP,Ma...   
EVE Online                     Space,Massively Multiplayer,Sci-fi,Sandbox,MMO...   

                                                                    game_details   
name                                                                               
DOOM                           Single-player,Multi-player,Co-op,Steam Achieve...  \
PLAYERUNKNOWN'S BATTLEGROUNDS             Multi-player,Online Multi-Player,Stats   
BATTLETECH                     Single-player,Multi-player,Online Multi-Player...   
DayZ                           Multi-player,Online Multi-Player,Steam Worksho...   
EVE Online                     Multi-player,Online Multi-Player,MMO,Co-op,Onl...   

                                                                           genre   
name                                                                               
DOOM                                                                      Action  \
PLAYERUNKNOWN'S BATTLEGROUNDS             Action,Adventure,Massively Multiplayer   
BATTLETECH                                             Action,Adventure,Strategy   
DayZ                                      Action,Adventure,Massively Multiplayer   
EVE Online                     Action,Free to Play,Massively Multiplayer,RPG,...   

                               has_setting  published_by_developer   
name                                                                 
DOOM                                 False                   False  \
PLAYERUNKNOWN'S BATTLEGROUNDS        False                    True   
BATTLETECH                           False                   False   
DayZ                                 False                    True   
EVE Online                           False                    True   

                               multiple_languages  
name                                               
DOOM                                         True  
PLAYERUNKNOWN'S BATTLEGROUNDS                True  
BATTLETECH                                   True  
DayZ                                         True  
EVE Online                                   True

In [43]:
multiple_languages = dataframe_clean[dataframe_clean["multiple_languages"]]
multiple_languages.shape

(13610, 8)

**13610 rows respond with True for column "multiple_languages".**

**5. Creating multiple columns regarding unique tags, details and genres. They will store boolean data type regarding the string they were extracted from.**

**Listing and counting the tags from column "popular_tags".**

In [44]:
all_tags = ",".join(dataframe_clean["popular_tags"])
tag_list = all_tags.split(',')
tag_counts = Counter(tag_list)
most_popular_tags = tag_counts.most_common()
print("Most popular tags:")
for tag, count in most_popular_tags:
    if count > 300:
        print(f"{tag}: {count}")

Most popular tags:
Indie: 16669
Action: 11793
Adventure: 9944
Casual: 8822
Simulation: 6183
Singleplayer: 5975
Strategy: 5794
Early Access: 4959
RPG: 4770
Great Soundtrack: 2824
Multiplayer: 2752
2D: 2695
Atmospheric: 2673
Puzzle: 2581
VR: 2043
Free to Play: 1898
Story Rich: 1871
Difficult: 1823
Violent: 1720
Horror: 1693
Anime: 1662
Pixel Graphics: 1552
Funny: 1521
Shooter: 1499
Platformer: 1491
Gore: 1467
Open World: 1455
First-Person: 1454
Sci-fi: 1447
Female Protagonist: 1439
Fantasy: 1424
Co-op: 1417
Sports: 1302
Retro: 1295
FPS: 1210
Arcade: 1200
Nudity: 1123
Survival: 1117
Family Friendly: 1088
Racing: 1058
Visual Novel: 1016
Comedy: 1005
Sandbox: 1002
Classic: 954
Cute: 948
Point & Click: 936
Massively Multiplayer: 931
Sexual Content: 912
Turn-Based: 852
Exploration: 850
Masterpiece: 849
Replay Value: 776
Space: 768
Psychological Horror: 746
Third Person: 731
Relaxing: 730
Local Multiplayer: 638
Mystery: 615
Colorful: 611
Tactical: 602
Controller: 589
Fast-Paced: 588
Zombies: 5

**Creating new columns according to uniqe tags.**

In [45]:
the_most_popular_tags = [(tag, count) for tag, count in tag_counts.items() if count > 300]
for tag, count in the_most_popular_tags:
    dataframe_clean[tag] = dataframe_clean["popular_tags"].apply(lambda x: tag in x)

In [46]:
dataframe_clean.head()

all_reviews   
name                                                                               
DOOM                           Very Positive,(42,550),- 92% of the 42,550 use...  \
PLAYERUNKNOWN'S BATTLEGROUNDS  Mixed,(836,608),- 49% of the 836,608 user revi...   
BATTLETECH                     Mostly Positive,(7,030),- 71% of the 7,030 use...   
DayZ                           Mixed,(167,115),- 61% of the 167,115 user revi...   
EVE Online                     Mostly Positive,(11,481),- 74% of the 11,481 u...   

                               release_date   
name                                          
DOOM                                   2016  \
PLAYERUNKNOWN'S BATTLEGROUNDS          2017   
BATTLETECH                             2018   
DayZ                                   2018   
EVE Online                             2003   

                                                                    popular_tags   
name                                                                               
DOOM                           FPS,Gore,Action,Demons,Shooter,First-Person,Gr...  \
PLAYERUNKNOWN'S BATTLEGROUNDS  Survival,Shooter,Multiplayer,Battle Royale,PvP...   
BATTLETECH                     Mechs,Strategy,Turn-Based,Turn-Based Tactics,S...   
DayZ                           Survival,Zombies,Open World,Multiplayer,PvP,Ma...   
EVE Online                     Space,Massively Multiplayer,Sci-fi,Sandbox,MMO...   

                                                                    game_details   
name                                                                               
DOOM                           Single-player,Multi-player,Co-op,Steam Achieve...  \
PLAYERUNKNOWN'S BATTLEGROUNDS             Multi-player,Online Multi-Player,Stats   
BATTLETECH                     Single-player,Multi-player,Online Multi-Player...   
DayZ                           Multi-player,Online Multi-Player,Steam Worksho...   
EVE Online                     Multi-player,Online Multi-Player,MMO,Co-op,Onl...   

                                                                           genre   
name                                                                               
DOOM                                                                      Action  \
PLAYERUNKNOWN'S BATTLEGROUNDS             Action,Adventure,Massively Multiplayer   
BATTLETECH                                             Action,Adventure,Strategy   
DayZ                                      Action,Adventure,Massively Multiplayer   
EVE Online                     Action,Free to Play,Massively Multiplayer,RPG,...   

                               has_setting  published_by_developer   
name                                                                 
DOOM                                 False                   False  \
PLAYERUNKNOWN'S BATTLEGROUNDS        False                    True   
BATTLETECH                           False                   False   
DayZ                                 False                    True   
EVE Online                           False                    True   

                               multiple_languages    FPS   Gore  ...   
name                                                             ...   
DOOM                                         True   True   True  ...  \
PLAYERUNKNOWN'S BATTLEGROUNDS                True   True  False  ...   
BATTLETECH                                   True  False  False  ...   
DayZ                                         True   True  False  ...   
EVE Online                                   True  False  False  ...   

                               Relaxing  Replay Value  Education   
name                                                               
DOOM                              False         False      False  \
PLAYERUNKNOWN'S BATTLEGROUNDS     False         False      False   
BATTLETECH                        False         False      False   
DayZ                              False        

**Listing and counting the tags from column "game_details".**

In [47]:
all_details = ",".join(dataframe_clean["game_details"])
details_list = all_details.split(',')
details_counts = Counter(details_list)
most_popular_details = details_counts.most_common()
print("Most popular details:")
for detail, count in most_popular_details:
    if count > 300:
        print(f"{detail}: {count}")

Most popular details:
Single-player: 24017
Steam Achievements: 14119
Steam Trading Cards: 8424
Steam Cloud: 8187
Full controller support: 6287
Downloadable Content: 5357
Multi-player: 5289
Partial Controller Support: 4443
Profile Features Limited 
									: 4304
Steam Leaderboards: 3758
Online Multi-Player: 2771
Co-op: 2463
Shared/Split Screen: 2130
Stats: 1848
Steam Workshop: 1461
Local Multi-Player: 1434
Online Co-op: 1301
Cross-Platform Multiplayer: 1259
Includes level editor: 1133
Local Co-op: 951
In-App Purchases: 869
Captions available: 785
Steam is learning about this game 
									: 666
MMO: 536


**Creating new columns according to uniqe details.**

In [48]:
the_most_popular_details = [(detail, count) for detail, count in details_counts.items() if count > 500]
for detail, count in the_most_popular_details:
    new_column_name = f"detail_{detail}"
    dataframe_clean[new_column_name] = dataframe_clean["game_details"].apply(lambda x: detail in x)

In [49]:
dataframe_clean.head()

all_reviews   
name                                                                               
DOOM                           Very Positive,(42,550),- 92% of the 42,550 use...  \
PLAYERUNKNOWN'S BATTLEGROUNDS  Mixed,(836,608),- 49% of the 836,608 user revi...   
BATTLETECH                     Mostly Positive,(7,030),- 71% of the 7,030 use...   
DayZ                           Mixed,(167,115),- 61% of the 167,115 user revi...   
EVE Online                     Mostly Positive,(11,481),- 74% of the 11,481 u...   

                               release_date   
name                                          
DOOM                                   2016  \
PLAYERUNKNOWN'S BATTLEGROUNDS          2017   
BATTLETECH                             2018   
DayZ                                   2018   
EVE Online                             2003   

                                                                    popular_tags   
name                                                                               
DOOM                           FPS,Gore,Action,Demons,Shooter,First-Person,Gr...  \
PLAYERUNKNOWN'S BATTLEGROUNDS  Survival,Shooter,Multiplayer,Battle Royale,PvP...   
BATTLETECH                     Mechs,Strategy,Turn-Based,Turn-Based Tactics,S...   
DayZ                           Survival,Zombies,Open World,Multiplayer,PvP,Ma...   
EVE Online                     Space,Massively Multiplayer,Sci-fi,Sandbox,MMO...   

                                                                    game_details   
name                                                                               
DOOM                           Single-player,Multi-player,Co-op,Steam Achieve...  \
PLAYERUNKNOWN'S BATTLEGROUNDS             Multi-player,Online Multi-Player,Stats   
BATTLETECH                     Single-player,Multi-player,Online Multi-Player...   
DayZ                           Multi-player,Online Multi-Player,Steam Worksho...   
EVE Online                     Multi-player,Online Multi-Player,MMO,Co-op,Onl...   

                                                                           genre   
name                                                                               
DOOM                                                                      Action  \
PLAYERUNKNOWN'S BATTLEGROUNDS             Action,Adventure,Massively Multiplayer   
BATTLETECH                                             Action,Adventure,Strategy   
DayZ                                      Action,Adventure,Massively Multiplayer   
EVE Online                     Action,Free to Play,Massively Multiplayer,RPG,...   

                               has_setting  published_by_developer   
name                                                                 
DOOM                                 False                   False  \
PLAYERUNKNOWN'S BATTLEGROUNDS        False                    True   
BATTLETECH                           False                   False   
DayZ                                 False                    True   
EVE Online                           False                    True   

                               multiple_languages    FPS   Gore  ...   
name                                                             ...   
DOOM                                         True   True   True  ...  \
PLAYERUNKNOWN'S BATTLEGROUNDS                True   True  False  ...   
BATTLETECH                                   True  False  False  ...   
DayZ                                         True   True  False  ...   
EVE Online                                   True  False  False  ...   

                               detail_Local Co-op  detail_Shared/Split Screen   
name                                                                            
DOOM                                        False                       False  \
PLAYERUNKNOWN'S BATTLEGROUNDS               False                       False   
BATTLETECH                                  False              

**Listing and counting the tags from column "genre".**

In [50]:
all_genres = ",".join(dataframe_clean["genre"])
genre_list = all_genres.split(",")
genre_counts = Counter(genre_list)
most_popular_genres = [(genre, count) for genre, count in genre_counts.items() if count > 300]
print("Most popular genres:")
for genre, count in most_popular_genres:
    print(f"{genre}: {count}")

Most popular genres:
Action: 11064
Adventure: 8826
Massively Multiplayer: 809
Strategy: 5258
Free to Play: 1700
RPG: 4417
Indie: 15950
Early Access: 2193
Simulation: 5705
Racing: 994
Casual: 7982
Sports: 1224


**Creating new columns according to uniqe genres.**

In [51]:
the_most_popular_genres = [(genre, count) for genre, count in genre_counts.items() if count > 500]
for genre, count in the_most_popular_genres:
    new_column_name = f"genre_{genre}"
    dataframe_clean[new_column_name] = dataframe_clean["genre"].apply(lambda x: genre in x)

In [52]:
dataframe_clean.head()

all_reviews   
name                                                                               
DOOM                           Very Positive,(42,550),- 92% of the 42,550 use...  \
PLAYERUNKNOWN'S BATTLEGROUNDS  Mixed,(836,608),- 49% of the 836,608 user revi...   
BATTLETECH                     Mostly Positive,(7,030),- 71% of the 7,030 use...   
DayZ                           Mixed,(167,115),- 61% of the 167,115 user revi...   
EVE Online                     Mostly Positive,(11,481),- 74% of the 11,481 u...   

                               release_date   
name                                          
DOOM                                   2016  \
PLAYERUNKNOWN'S BATTLEGROUNDS          2017   
BATTLETECH                             2018   
DayZ                                   2018   
EVE Online                             2003   

                                                                    popular_tags   
name                                                                               
DOOM                           FPS,Gore,Action,Demons,Shooter,First-Person,Gr...  \
PLAYERUNKNOWN'S BATTLEGROUNDS  Survival,Shooter,Multiplayer,Battle Royale,PvP...   
BATTLETECH                     Mechs,Strategy,Turn-Based,Turn-Based Tactics,S...   
DayZ                           Survival,Zombies,Open World,Multiplayer,PvP,Ma...   
EVE Online                     Space,Massively Multiplayer,Sci-fi,Sandbox,MMO...   

                                                                    game_details   
name                                                                               
DOOM                           Single-player,Multi-player,Co-op,Steam Achieve...  \
PLAYERUNKNOWN'S BATTLEGROUNDS             Multi-player,Online Multi-Player,Stats   
BATTLETECH                     Single-player,Multi-player,Online Multi-Player...   
DayZ                           Multi-player,Online Multi-Player,Steam Worksho...   
EVE Online                     Multi-player,Online Multi-Player,MMO,Co-op,Onl...   

                                                                           genre   
name                                                                               
DOOM                                                                      Action  \
PLAYERUNKNOWN'S BATTLEGROUNDS             Action,Adventure,Massively Multiplayer   
BATTLETECH                                             Action,Adventure,Strategy   
DayZ                                      Action,Adventure,Massively Multiplayer   
EVE Online                     Action,Free to Play,Massively Multiplayer,RPG,...   

                               has_setting  published_by_developer   
name                                                                 
DOOM                                 False                   False  \
PLAYERUNKNOWN'S BATTLEGROUNDS        False                    True   
BATTLETECH                           False                   False   
DayZ                                 False                    True   
EVE Online                           False                    True   

                               multiple_languages    FPS   Gore  ...   
name                                                             ...   
DOOM                                         True   True   True  ...  \
PLAYERUNKNOWN'S BATTLEGROUNDS                True   True  False  ...   
BATTLETECH                                   True  False  False  ...   
DayZ                                         True   True  False  ...   
EVE Online                                   True  False  False  ...   

                               genre_Massively Multiplayer  genre_Strategy   
name                                                                         
DOOM                                                 False           False  \
PLAYERUNKNOWN'S BATTLEGROUNDS                         True           False   
BATTLETECH                                           False            True 

**Deleting the columns "genre", "popular_tags" and "game_details" as I no longer need them.**

In [53]:
dataframe_clean = dataframe_clean.drop(["genre", "popular_tags", "game_details"], axis=1)
dataframe_clean.head()

all_reviews   
name                                                                               
DOOM                           Very Positive,(42,550),- 92% of the 42,550 use...  \
PLAYERUNKNOWN'S BATTLEGROUNDS  Mixed,(836,608),- 49% of the 836,608 user revi...   
BATTLETECH                     Mostly Positive,(7,030),- 71% of the 7,030 use...   
DayZ                           Mixed,(167,115),- 61% of the 167,115 user revi...   
EVE Online                     Mostly Positive,(11,481),- 74% of the 11,481 u...   

                               release_date  has_setting   
name                                                       
DOOM                                   2016        False  \
PLAYERUNKNOWN'S BATTLEGROUNDS          2017        False   
BATTLETECH                             2018        False   
DayZ                                   2018        False   
EVE Online                             2003        False   

                               published_by_developer  multiple_languages   
name                                                                        
DOOM                                            False                True  \
PLAYERUNKNOWN'S BATTLEGROUNDS                    True                True   
BATTLETECH                                      False                True   
DayZ                                             True                True   
EVE Online                                       True                True   

                                 FPS   Gore  Action  Shooter  First-Person   
name                                                                         
DOOM                            True   True    True     True          True  \
PLAYERUNKNOWN'S BATTLEGROUNDS   True  False    True     True          True   
BATTLETECH                     False  False    True    False         False   
DayZ                            True  False    True     True         False   
EVE Online                     False  False    True    False         False   

                               ...  genre_Massively Multiplayer   
name                           ...                                
DOOM                           ...                        False  \
PLAYERUNKNOWN'S BATTLEGROUNDS  ...                         True   
BATTLETECH                     ...                        False   
DayZ                           ...                         True   
EVE Online                     ...                         True   

                               genre_Strategy  genre_Free to Play  genre_RPG   
name                                                                           
DOOM                                    False               False      False  \
PLAYERUNKNOWN'S BATTLEGROUNDS           False               False      False   
BATTLETECH                               True               False      False   
DayZ                                    False               False      False   
EVE Online                               True                True       True   

                               genre_Indie  genre_Early Access   
name                                                             
DOOM                                 False               False  \
PLAYERUNKNOWN'S BATTLEGROUNDS        False               False   
BATTLETECH                           False               False   
DayZ                                 False               False   
EVE Online                           False               False   

                               genre_Simulation  genre_Racing  genre_Casual   
name                                                                          
DOOM                                      False         False         False  \
PLAYERUNKNOWN'S BATTLEGROUNDS             False         False         False   
BATTLETECH                                False         False         False   
DayZ                                      False         False         False   
EVE 

In [54]:
dataframe_clean.shape

(25697, 149)

**6. Creating columns "total_reviews" and "positive_reviews_share" regarding the column "all_reviews". They will contain int and float data type accordingly.**

**Checking if all rows have a sufficient string to extract data. Deleting those which do not have.**

In [55]:
indices = dataframe_clean[dataframe_clean["all_reviews"].str.contains("Need more", na=False)].index
dataframe_clean = dataframe_clean.drop(indices)

In [56]:
dataframe_clean.shape

(16344, 149)

**9353 rows were deleted as they lacked required data.**

In [57]:
# function extracts first and second number from a string and returns them as int and float data type accordingly
def extract_numbers(string):
    total_reviews = None
    positive_reviews_share = None
    match_total = re.search(r"\(([\d,]+)\)", string)
    match_positive = re.search(r"(\d+)%", string)
    if match_total:
        total_reviews = int(match_total.group(1).replace(",", ""))
    if match_positive:
        positive_reviews_share = float(match_positive.group(1)) / 100
    return total_reviews, positive_reviews_share

In [58]:
dataframe_clean[["total_reviews", 
                 "positive_reviews_share"]] = dataframe_clean["all_reviews"].apply(lambda x: pd.Series(extract_numbers(x)))

**"positive_reviews_share" is gona be the target of the dataset.**

**Deleting the column "all_reviews" as data was exctracted already.**

In [59]:
dataframe_clean = dataframe_clean.drop(["all_reviews"], axis=1)

**Checking on how dataset looks like now.**

In [60]:
dataframe_clean

release_date  has_setting   
name                                                       
DOOM                                   2016        False  \
PLAYERUNKNOWN'S BATTLEGROUNDS          2017        False   
BATTLETECH                             2018        False   
DayZ                                   2018        False   
EVE Online                             2003        False   
...                                     ...          ...   
Grabity                                2018        False   
Detached: Non-VR Edition               2018        False   
A Room Beyond                          2017        False   
Chasing the Stars                      2019        False   
Nanoui                                 2018        False   

                               published_by_developer  multiple_languages   
name                                                                        
DOOM                                            False                True  \
PLAYERUNKNOWN'S BATTLEGROUNDS                    True                True   
BATTLETECH                                      False                True   
DayZ                                             True                True   
EVE Online                                       True                True   
...                                               ...                 ...   
Grabity                                          True                True   
Detached: Non-VR Edition                         True                True   
A Room Beyond                                    True                True   
Chasing the Stars                                True               False   
Nanoui                                           True               False   

                                 FPS   Gore  Action  Shooter  First-Person   
name                                                                         
DOOM                            True   True    True     True          True  \
PLAYERUNKNOWN'S BATTLEGROUNDS   True  False    True     True          True   
BATTLETECH                     False  False    True    False         False   
DayZ                            True  False    True     True         False   
EVE Online                     False  False    True    False         False   
...                              ...    ...     ...      ...           ...   
Grabity                        False  False    True    False         False   
Detached: Non-VR Edition       False  False   False    False         False   
A Room Beyond                  False  False   False    False         False   
Chasing the Stars              False  False   False    False         False   
Nanoui                         False  False    True    False         False   

                               Great Soundtrack  ...  genre_Free to Play   
name                                             ...                       
DOOM                                       True  ...               False  \
PLAYERUNKNOWN'S BATTLEGROUNDS             False  ...               False   
BATTLETECH                                 True  ...               False   
DayZ                                      False  ...               False   
EVE Online                                False  ...                True   
...                                         ...  ...                 ...   
Grabity                                   False  ...               False   
Detached: Non-VR Edition                   True  ...               False   
A Room Beyond                             False  ...               False   
Chasing the Stars                         False  ...               False   
Nanoui                                     True  ...               False   

                               genre_RPG  genre_Indie  genre_Early Access   
name                                                                        
DOOM                               False        False               False  \
PLAYERUNKNOWN

# 4. Saving the reshaped dataset

In [61]:
dataframe_clean.to_csv("Datasets/steam_games_cleaned.csv", index=False)